lib/internal/modules/esm/hooks.js

```diff
@@ -105,7 +104,39 @@ function defineImportAssertionAlias(context) {
 
 // [2] `validate...()`s throw the wrong error
 
-class Hooks {
+/**
+ * @typedef {{ format: ModuleFormat, source: ModuleSource }} LoadResult
+ */
+
+/**
+ * @typedef {{ format: ModuleFormat, url: string, importAttributes: Record<string, string> }} ResolveResult
+ */
+
+/**
+ * Interface for classes that implement asynchronous loader hooks that can be attached to the ModuleLoader
+ * via `ModuleLoader.#setAsyncLoaderHooks()`.
+ * @typedef {object} AsyncLoaderHooks
```

Comment:
Nit: the reason for the “customization“ and “customization hooks“ names is to avoid confusion where some might think that these hooks only affect loading, when they also impact resolution and other things. It hasn’t been a priority to update the code, but as long as we’re adding new code and defining new names, I think the “customization“ name is more informative and correct.

```suggestion
 * @typedef {object} AsyncCustomizationHooks
```

And elsewhere that the code uses “loader hook“ that this branch touches.

Reply:
In that case should `AsyncLoaderHooksOnLoaderHookWorker` be `AsyncCustomizationHooksOnCustomizationHookWorker`? That feels rather lengthy..

I feel that "loader hooks" contain already ample information about it: it's `loader`, which implies it's not just "loading" phase but everything you'd see in `loader.js` and the `ModuleLoader` class. I can see that "load" would be ambuiguous, but "loader" isn't, and "hooks" implies customizations (or the hooks aren't necessarily for customizations - they can also be for observations).

Reply:
It's just a nit, do with it as you will. `AsyncCustomizationHooksOnCustomizationHookWorker` is the same number of words as with `Loader`, it's just a longer word. It's a mouthful in either version. Perhaps just `AsyncOffThreadCustomizationHooks`?

Another name instead of “customization hooks“ is “module hooks“ which is just as descriptive and the same number of letters as “loader hooks”.

Reply:
I appreciate the brainstorm ;) though to me "foo hooks" means "it's hooking into foo/hook placed in foo", so I'd have some questions when I see "module hooks" - is it hooking into the code of user modules/hook placed in the modules? While it can be used in such a way that inject code into user modules, that is more of the output, instead of the process. "loader hooks" seem more intuitive because, there's a `ModuleLoader` in `loader.js` next to it, which either `new AsyncLoaderHooksProxiedToLoaderHookWorker` itself, or get a `new AsyncLoaderHooksOnLoaderHookWorker` passed into it, and the two classes have methods that are hooking into the methods of `ModuleLoader` with the same name, so that makes a lot of sense.

(I thought about just calling it `async hooks` but then unfortunately that combination might be a bit too well known as something completely unrelated in Node.js..)